README.md


# quadrupen: Sparsity by Worst-Case Quadratic Penalties

## Description

Fits classical sparse regression models with efficient active set
algorithms by solving quadratic problems. Also provides a few methods
for model selection purpose (cross-validation, stability selection).

## Installation


```r
devtools::install_github("jchiquet/quadrupen").
```

Many examples and demos can be found in the `inst` repositories.  Here is a more illustraive one: 

## Example: structured penalized regression

This example is extracted form chapter 1 of my [habilitation](https://tel.archives-ouvertes.fr/tel-01288976/). You can get more insight by reading pages 65-66, 70-71. 

Ok first load the package


```r
library(quadrupen)
```

### A toy data set advocating for structured regularization

See pages 29--30 in my [habilitation](https://tel.archives-ouvertes.fr/tel-01288976/). COde for additional funciton is given by the end of the present document (Appendix).



We draw data from linear regression where the regression parameters are defined groupwise. The corresponding regression correlated according to the same pattern.

```r
n <- 200
p <- 192

## model settings: block wise
mu <- 0
group <- c(p/4,p/8,p/4,p/8,p/4)
label <- factor(rep(paste("group", 1:5), group))
beta <- rep(c(0.25,1,-0.25,-1,0.25), group)
x <- rPred.block(n, p, sizes = group, rho=c(0.25, 0.75, 0.25, 0.75, 0.25))
```

Indeed, the correlation structure between the regressor exhibitis a strong pattern:
![plot of chunk unnamed-chunk-3](figure/unnamed-chunk-3-1.png)

We draw the reponse variable by fixing the variance of the noise ratio to get an $R^2$ equal to $0.8$.

```r
dat <- rlm(x,beta,mu,r2=0.8)
```

```
## Warning in rnorm(n) * sigma: Recycling array of length 1 in vector-array arithmetic is deprecated.
##   Use c() or as.vector() instead.
```

```r
y <- dat$y
sigma <- dat$sigma
```

### Regularization without prior knowledge

No we try the available penalized regression method in `quadrupen` to fit the regularization path to this data set

#### Ridge regression  ($\ell_2$ penalty)


```r
plot(ridge(x,y, intercept=FALSE), label=label)
```

![plot of chunk ridge_nostruct](figure/ridge_nostruct-1.png)

#### Lasso  ($\ell_1$ penalty)


```r
plot(lasso(x,y, intercept=FALSE), label=label)
```

![plot of chunk lasso_nostruct](figure/lasso_nostruct-1.png)

#### Elastic-net  ($\ell_1+\ell_2$ penalty)


```r
plot(elastic.net(x,y, lambda2=1, intercept=FALSE), label=label)
```

![plot of chunk enet_nostruct](figure/enet_nostruct-1.png)

#### Bounded regression ($\ell_\infty$ penalty)


```r
plot(bounded.reg(x,y, lambda2=0, intercept=FALSE), label=label)
```

![plot of chunk breg_nostruct](figure/breg_nostruct-1.png)

#### Ridge + Bounded regression ($\ell_\infty+\ell_2$ penalty)


```r
plot(bounded.reg(x,y, lambda2=5, intercept=FALSE), label=label)
```

![plot of chunk bregridge_nostruct](figure/bregridge_nostruct-1.png)

### Regularization with prior knowledge

Now let use define the graph associated with the groups of the regressor and compute the Graph Laplacian. We add a small value on the diagonal ton ensure strict positive definiteness (in he futre, this matrix should preferentially be defined with a graph).


```r
A <- bdiag(lapply(group, function(s) matrix(1,s,s))) ; diag(A) <- 0
L <- -A; diag(L) <- colSums(A) + 1e-2
```

Now, we run all the method having a ridge-like regularization by replacing the ridge penalty $\| \code \|_2^2$ with $\| \code \|_{\mathbf{L}}^2$ to enforce some structure in the regularization: the solution paths look a lot more convincing.

#### Ridge regression  ($\ell_2$ penalty)


```r
plot(ridge(x,y, lambda2 = 10^seq(4,-1,len=100), struct = L, intercept=FALSE), label=label)
```

![plot of chunk ridge_struct](figure/ridge_struct-1.png)


#### Elastic-net  ($\ell_1+\ell_2$ penalty)


```r
plot(elastic.net(x,y, struct = L, lambda2=1, intercept=FALSE), label=label)
```

![plot of chunk enet_struct](figure/enet_struct-1.png)

#### Ridge + Bounded regression ($\ell_\infty+\ell_2$ penalty)


```r
plot(bounded.reg(x,y, struct = L, lambda2=5, intercept=FALSE), label=label)
```

![plot of chunk bregridge_struct](figure/bregridge_struct-1.png)

## Appendix: functions for data generation


```r
chol.uniform <- function(p,rho) {

  q <- 0     ## sum_k^(i-1) c.k^2
  cii <- 1   ## current value of the diagonal
  c.i <- rho ## current value of the off-diagonal term
  Cii <- cii ## vector of diagonal terms
  C.i <- c.i ## vector of off-diagonal terms
  for (i in 2:p) {
    q <- q + c.i^2
    cii <- sqrt(1-q)
    c.i <- (rho-q)/cii
    Cii <- c(Cii,cii)
    C.i <- c(C.i,c.i)
  }
  return(list(cii=Cii, c.i=C.i[-p]))
}

rlm <- function(x,beta,mu=0,r2=NULL,sigma=1) {
  n <- nrow(x)
  if (!is.null(r2))
      sigma  <- sqrt((1-r2)/r2 * t(beta) %*% cov(x) %*% beta)
  epsilon <- rnorm(n) * sigma
  y <- mu + x %*% beta + epsilon
  r2 <- 1 - sum(epsilon^2) / sum((y-mean(y))^2)
  return(list(y=y,sigma=sigma))
}

## Blockwise structure
## the length of rho determine the number of blocs
## each group must be > 1 individual
rPred.block <- function(n, p, sizes=rmultinom(1,p,rep(p/K,K)), rho=rep(0.75,4)) {

  K <- length(rho)
  stopifnot(sum(sizes) == p | length(rho) != length(sizes))
  Cs <- lapply(1:K, function(k) chol.uniform(sizes[k],rho[k]))
  
  rmv <- function() {
      return(
      unlist(lapply(1:K, function(k) {
        z <- rnorm(sizes[k])
        Cs[[k]]$cii*z + c(0,cumsum(z[-sizes[k]]*Cs[[k]]$c.i))
      }))
      )
  }
  return(t(replicate(n, rmv(), simplify=TRUE)))
}
```




NEWS
NEWS/Changelog

0.3-0	(2013-11-xx)
    Major:
    +	added structured ridge regression (regression penalized by structered l2 norm)
    +   added corresponding functionalies for cross-validation and stability path
    Minor:
    +   removed the not very useful 'reverse' argument for plotting    
    +   added demo for lasso, ridge and use of the criteria function    

0.2-4	(2013-11-01)
    +   added a 'lasso' function, simple wrapper to the elastic-net 'function'
    +   added computation of degrees of freedom (for elastic net and bounded regression)
    +   added a method to compute penalized criteria (BIC/AIC) of a quadrupen fit, with plot

0.2-3	(2013-08-26)
    +   added back the 'normalize' parameter
    +   standardization is performed within the C++ code
    +   use of sparse conversion from Matrix to Armadillo
    +   corrected bug with the 'intercept' and 'residuals' components of the quadrupen class
    +   added more tests in the inst directory
    +   correction in the documentation
    +   added r.squared to the quadrupen class

0.2-2	(2013-04-08)
    +   minor fix to comply with recent ggplot2 updates.

0.2-1	(2013-02-27)
    +   minor fix to pass CRAN check on Windows operating systems.

0.2-0	(2013-02-26)
    Major:
    +	added bounded regression (regression penalized by infinity norm + structered l2 norm)
    +   added corresponding functionalies for cross-validation and stability path
    Minor:
    +   corrected wrong annotations of the stability path (PFER)
    +   handled normalization internally ('normalize' is no longer a parameter)
    +   more simple internal handling of penscales and correction of the rescaling of the intercept
    +   better use of multicore features
    +   handled runtime error exception in RcppArmadillo when the system is singular (end of the solution path)
        A consequence is quadrupen is less likely to crash due to user's "bad" parametrization
    +   simplification of the C++ code, bugs corrected, probably new ones added :-'(
    +   added 'examples' and 'tests' directories

0.1-0	(2012-10-09)
    +	first build: structured elastic-net with (weighted) quadratic loss, cross-validation and stability selection methods.


R/stability.R
##' Stability selection for a quadrupen fit.
##'
##' Compute the stability path of a (possibly randomized) fitting
##' procedure as introduced by Meinshausen and Buhlmann (2010).
##'
##' @param x matrix of features, possibly sparsely encoded
##' (experimental). Do NOT include intercept.
##'
##' @param y response vector.
##'
##' @param penalty a string for the fitting procedure used for
##' cross-validation. Either \code{\link{elastic.net}},
##' \code{\link{lasso}} or \code{"bounded.reg"}. this is of course
##' useless for \code{"ridge"} regression.
##'
##' @param subsamples integer indicating the number of subsamplings
##' used to estimate the selection probabilities. Default is 100.
##'
##' @param sample.size integer indicating the size of each subsamples.
##' Default is \code{floor(n/2)}.
##'
##' @param randomize Should a randomized version of the fitting
##' procedure by used? Default is \code{TRUE}. See details below.
##'
##' @param weakness Coefficient used for randomizing. Default is
##' \code{0.5}. Ignored when \code{randomized} is \code{FALSE}. See
##' details below.
##'
##' @param folds list with \code{subsamples} entries with vectors
##' describing the folds to use for the stability procedure. By
##' default, the folds are randomly sampled with the specified
##' \code{subsamples} argument.
##'
##' @param verbose logical; indicates if the progression should be
##' displayed. Default is \code{TRUE}.
##'
##' @param mc.cores the number of cores to use. The default uses all
##' the cores available.
##'
##' @param ... additional parameters to overwrite the defaults of the
##' fitting procedure. See the corresponding documentation
##' (\code{\link{elastic.net}},\code{\link{lasso}} or
##' \code{\link{bounded.reg}})
##'
##' @return An object of class \code{\linkS4class{stability.path}}.
##'
##' @note When \code{randomized = TRUE}, the \code{penscale} argument
##' that weights the penalty tuned by \eqn{\lambda_1}{lambda1} is
##' perturbed (divided) for each subsample by a random variable
##' uniformly distributed on
##' \if{latex}{\eqn{[\alpha,1]}}\if{html}{[&#945;,1]}\if{text}{\eqn{[alpha,1]}},
##' where
##' \if{latex}{\eqn{\alpha}}\if{html}{&#945;}\if{text}{\eqn{alpha}} is
##' the weakness parameter.
##'
##' If the user runs the fitting method with option
##' \code{'bulletproof'} set to \code{FALSE}, the algorithm may stop
##' at an early stage of the path. Early stops of the underlying
##' fitting function are handled internally, in the following way: we
##' chose to simply skip the results associated with such runs, in
##' order not to bias the stability selection procedure. If it occurs
##' too often, a warning is sent to the user, in which case you should
##' reconsider the grid of \code{lambda1} for stability selection. If
##' \code{bulletproof} is \code{TRUE} (the default), there is nothing
##' to worry about, except a possible slow down when any switching to
##' the proximal algorithm is required.
##'
##' @references N. Meinshausen and P. Buhlmann (2010). Stability
##' Selection, JRSS(B).
##'
##' @seealso \code{\linkS4class{stability.path}} and
##' \code{\link{plot.stability.path}}.
##'
##' @examples \dontrun{
##' ## Simulating multivariate Gaussian with blockwise correlation
##' ## and piecewise constant vector of parameters
##' beta <- rep(c(0,1,0,-1,0), c(25,10,25,10,25))
##' Soo  <- matrix(0.75,25,25) ## bloc correlation between zero variables
##' Sww  <- matrix(0.75,10,10) ## bloc correlation between active variables
##' Sigma <- bdiag(Soo,Sww,Soo,Sww,Soo) + 0.2
##' diag(Sigma) <- 1
##' n <- 100
##' x <- as.matrix(matrix(rnorm(95*n),n,95) %*% chol(Sigma))
##' y <- 10 + x %*% beta + rnorm(n,0,10)
##'
##' ## Build a vector of label for true nonzeros
##' labels <- rep("irrelevant", length(beta))
##' labels[beta != 0] <- c("relevant")
##' labels <- factor(labels, ordered=TRUE, levels=c("relevant","irrelevant"))
##'
##' ## Call to stability selection function, 200 subsampling
##' stab <- stability(x,y, subsamples=200, lambda2=1, min.ratio=1e-2)
##' ## Recover the selected variables for a given cutoff
##' ## and per-family error rate, without producing any plot
##' stabpath <- plot(stab, cutoff=0.75, PFER=1, plot=FALSE)
##'
##' cat("\nFalse positives for the randomized Elastic-net with stability selection: ",
##'      sum(labels[stabpath$selected] != "relevant"))
##' cat("\nDONE.\n")
##'}
##' @keywords models, regression
##'
##' @name stability
##' @aliases stability
##' @rdname stability
##'
##' @export
stability <- function(x,
                      y,
                      penalty = c("elastic.net", "lasso", "bounded.reg"),
                      subsamples  = 100,
                      sample.size = floor(n/2),
                      randomize   = TRUE,
                      weakness    = 0.5,
                      verbose     = TRUE,
                      folds       = replicate(subsamples, sample(1:nrow(x), sample.size), simplify=FALSE),
                      mc.cores    = detectCores(),
                      ...) {

  ## =============================================================
  ## INITIALIZATION & PARAMETERS RECOVERY
  if (Sys.info()[['sysname']] == "Windows") {
    warning("\nWindows does not support fork, enforcing mc.cores to '1'.")
    mc.cores <- 1
  }
  penalty <- match.arg(penalty)
  fit.func <- switch(penalty,
                        "elastic.net" = elastic.net,
                        "lasso"       = lasso      , 
                        "bounded.reg" = bounded.reg)
  get.lambda1 <- switch(penalty,
                        "elastic.net" = get.lambda1.l1,
                        "lasso"       = get.lambda1.l1,
                        "bounded.reg" = get.lambda1.li)
  if (penalty == "lasso") {lambda2 <- NULL}

  p <- ncol(x)
  n <- nrow(x)

  user <- list(...)
  defs <- default.args(penalty,n,p,user)
  args <- modifyList(defs, user)
  ## overwrite parameters irrelevant in and resampling context
  args$control$verbose <- 0
  args$max.feat  <- p
  args$checkargs <- FALSE
  ## Compute a grid of lambda1 (the smae for each fold)
  if (is.null(args$lambda1)) {
    input <- standardize(x,y,args$intercept,args$normalize,args$penscale)
    args$lambda1 <- get.lambda1(input$xty,args$nlambda1,args$min.ratio)
    rm(input)
  }
  nlambda1 <- length(args$lambda1)
  ## record the basic penscale value for randomizing
  penscale <- args$penscale
  ## Prepare blocs of sub samples to run jobs parallely
  blocs <- split(1:subsamples, 1:mc.cores)

  if (verbose) {
    cat(paste("\n\nSTABILITY SELECTION ",ifelse(randomize,"with","without")," randomization (weakness = ",weakness,")",sep=""))
    if(penalty == "lasso") {
      cat(paste("\nFitting procedure: ",penalty," with ",nlambda1,"-dimensional grid of lambda1.", sep=""))
    } else {
      cat(paste("\nFitting procedure: ",penalty," with lambda2 = ",args$lambda2," and an ",nlambda1,"-dimensional grid of lambda1.", sep=""))
    }
    
    cat("\nRunning",length(blocs),"jobs parallely (1 per core)")
    cat("\nApprox.", length(blocs[[1]]),"subsamplings for each job for a total of",subsamples)
  }

  ## function to run on each core
  bloc.stability <- function(subsets) {
    select <- Matrix(0,length(args$lambda1),p)
    subsamples.ok <- 0
    for (s in 1:length(subsets)) {
      if (randomize) {args$penscale <- penscale / runif(p,weakness,1)}
      active <- do.call(fit.func, c(list(x=x[folds[[subsets[s]]], ], y=y[folds[[subsets[s]]]]), args))@active.set
      if (nrow(active) == length(args$lambda1)) {
        subsamples.ok <- subsamples.ok + 1
        select <- select + active
      }
    }
    if (subsamples.ok < 0.5*length(subsets)) {
      cat("\nWarning: more than 50% of the run were discarded in that core due to early stops of the fitting procedure. You should consider largest 'min.ratio' or strongest 'lambda2'.")
    }
    return(select/(subsamples.ok*length(blocs)))
  }

  ## Now launch the B jobs...
  prob.bloc <- mclapply(blocs, bloc.stability, mc.cores=mc.cores)

  ## Construct the probability path
  path <- Matrix(0,nlambda1,p)
  for (b in 1:length(prob.bloc)) {
    path <- path + prob.bloc[[b]]
  }

  return(new("stability.path",
             probabilities = path        ,
             penalty       = penalty     ,
             naive         = args$naive  ,
             lambda1       = args$lambda1,
             lambda2       = args$lambda2,
             folds         = folds       ))

}
